File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    result = []
    for perm in permutations(dungeons,len(dungeons)):
        max_answer = 0
        temp_k = k
        for min_fatigue, minus in perm:
            if min_fatigue <= temp_k :
                temp_k-=minus
                max_answer += 1
        result.append(max_answer)
    answer = max(result)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {

    static int max = 0;
    static boolean[] visited;
    static int[][] dungeons;

    public int solution(int k, int[][] dungeons) {
        this.dungeons = dungeons;
        visited = new boolean[dungeons.length];

        dfs(k, 0);

        return max;
    }

    private void dfs(int k, int count) {
        max = Math.max(max, count);

        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;

                dfs(k - dungeons[i][1], count + 1);

                visited[i] = false;
            }
        }
    }
}